Allow importing tags from ImageStreams pointing to external registries

soltysh opened the pull request:
This is the initial version of improved importer, to address [card 514](https://trello.com/c/Y2MnDbaI/514-ability-to-define-imagestreamtag-pointing-to-a-different-image-registry-each).

@ncdc I know it's bit messy still, I'm working on getting it DRY-ed and most importantly fully tested, still I'd appreciate first raw pass, if this makes sense. The general flow is following:

1. if no tags are defined import all from default spec
2. if tags are defined import only the ones defined, both from default spec and from external repos

@pweil- fyi

ncdc commented:
cc @pweil- @miminar

ncdc commented on pkg/image/controller/controller.go:
```diff
@@ -39,35 +33,105 @@ const retryCount = 2
+	return c.done(stream, "", retryCount)
+}
+
+// getDefaultTags reads all tags from the stream's default registry,
```
s/registry/image repository/

ncdc commented on pkg/image/controller/controller.go:
```diff
@@ -39,35 +33,105 @@ const retryCount = 2
+	return tags, err, dockerregistry.IsRepositoryNotFound(err) || dockerregistry.IsRegistryNotFound(err)
+}
+
+// importDefaultTags imports tags from the stream's default registry
```
s/registry/image repository/

soltysh commented on pkg/image/registry/imagestream/strategy.go:
```diff
@@ -156,12 +156,6 @@ func (s Strategy) tagsChanged(old, stream *api.ImageStream) fielderrors.Validati
-				errs = append(errs, fielderrors.NewFieldInvalid(fmt.Sprintf("spec.tags[%s].from", tag), tagRef.From, err.Error()))
-				continue
-			}
-			api.AddTagEventToImageStream(stream, tag, *event)
```
@ncdc I'm still wondering about this change, though. Without it when importing tags I'm getting two events for external repository:
```
2.2	openshift/ruby-22-centos7	9 minutes ago	docker.io/openshift/ruby-22-centos7@sha256:5dc0...
					            9 minutes ago	openshift/ruby-22-centos7
```
First is created by above code which is not able to decompose the full spec at the time, because it wasn't imported yet, the second shows that actual import and since it was resolved in the mean time, creates a new tag event, because the event has changed. Is that ok?

smarterclayton replied:
No, it's not ok.  If you want to import something, we shouldn't create the status tag until it's imported.  If you don't want to import it, but instead to direct tag it, it shouldn't get a second tag.  If we want to import by default, then we need to bypass this.

smarterclayton replied:
Three use cases:

1. Import an entire registry (all tags)
2. Reference a different tag without metadata retrieval
3. Import a tag with metadata retrieval

1 and 3 are the most likely, neither 2 nor 3 can result in two stream updates.  Users need to be able to clearly define reference vs import

ncdc replied:
I think the description was confusing. With this PR, when you set spec.tags[2.2].from=openshift/ruby-22-centos7 (kind=DockerImage), you only get *1* image:

```yaml
  - items:
    - created: 2015-09-30T13:49:37Z
      dockerImageReference: openshift/ruby-20-centos7@sha256:509c5a47b689e1523388ff6c160b0ea1417f412794b4a9854277b272de7384d3
      image: sha256:509c5a47b689e1523388ff6c160b0ea1417f412794b4a9854277b272de7384d3
```

smarterclayton replied:
Ok, just wanted to be clear.  I think the challenge is that there will be
repos we can't import (because we don't have auth yet, or it's down) - if
that blocks an admin from directly tagging that's bad.  On the other hand,
trying to get the image metadata is really really important.  So there
still has to be a bypass import flow.

On Wed, Sep 30, 2015 at 3:59 PM, Andy Goldstein <notifications@github.com>
wrote:

> In pkg/image/registry/imagestream/strategy.go
> <https://github.com/openshift/origin/pull/4853#discussion_r40796195>:
>
> > @@ -156,12 +156,6 @@ func (s Strategy) tagsChanged(old, stream *api.ImageStream) fielderrors.Validati
> >  		}
> >
> >  		if tagRef.From.Kind == "DockerImage" && len(tagRef.From.Name) > 0 {
> > -			event, err := tagReferenceToTagEvent(stream, tagRef, "")
> > -			if err != nil {
> > -				errs = append(errs, fielderrors.NewFieldInvalid(fmt.Sprintf("spec.tags[%s].from", tag), tagRef.From, err.Error()))
> > -				continue
> > -			}
> > -			api.AddTagEventToImageStream(stream, tag, *event)
>
> I think the description was confusing. With this PR, when you set
> spec.tags[2.2].from=openshift/ruby-22-centos7 (kind=DockerImage), you only
> get *1* image:
>
>   - items:
>     - created: 2015-09-30T13:49:37Z
>       dockerImageReference: openshift/ruby-20-centos7@sha256:509c5a47b689e1523388ff6c160b0ea1417f412794b4a9854277b272de7384d3
>       image: sha256:509c5a47b689e1523388ff6c160b0ea1417f412794b4a9854277b272de7384d3
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/4853/files#r40796195>.
>

smarterclayton commented on pkg/image/controller/controller.go:
```diff
@@ -39,35 +33,105 @@ const retryCount = 2
+//    from default registry
+// 2. spec.DockerImageRepository + tags defined - import only the provided tags,
+//    for kind == DockerImage import underlying image, iow. exact tag (no provided means latest),
+//    for kind != DockerImage import tag from the default registry
```
Not really importing tag, instead it's "referencing" tag which we should be clear is not an import.

soltysh replied:
So here are the propositions for the boolean flag then:
- reference
- external
- foreign

 in the preference order of mine.

smarterclayton replied:
Let's assume "reference" for now.  Also, you mentioned in the trello card about adding this to the ImageStream.Spec, but I don't think that's necessary (reference tagging is always manual, we don't need to have an automated reference tag, just automated import tag).

soltysh replied:
I wasn't specific enough on the trello, the ImageStreamSpec will have additional ImportPolicy field, like @ncdc suggested that will by default will import all, and you can change it to import specified.

smarterclayton replied:
Why do we need import policy?  Specifying a DockerImageRepository in
ImageStreamSpec is an effective importPolicy of "all".  If you don't want
"all", create one tag per import.

On Thu, Oct 1, 2015 at 12:14 PM, Maciej Szulik <notifications@github.com>
wrote:

> In pkg/image/controller/controller.go
> <https://github.com/openshift/origin/pull/4853#discussion_r40898234>:
>
> > @@ -39,35 +33,105 @@ const retryCount = 2
> >  // the image stream is not modified (so it will be tried again later). If a permanent
> >  // failure occurs the image is marked with an annotation. The tags of the original spec image
> >  // are left as is (those are updated through status).
> > +// There are 3 use cases here:
> > +// 1. spec.DockerImageRepository defined without any tags results in all tags being imported
> > +//    from default registry
> > +// 2. spec.DockerImageRepository + tags defined - import only the provided tags,
> > +//    for kind == DockerImage import underlying image, iow. exact tag (no provided means latest),
> > +//    for kind != DockerImage import tag from the default registry
>
> I wasn't specific enough on the trello, the ImageStreamSpec will have
> additional ImportPolicy field, like @ncdc <https://github.com/ncdc>
> suggested that will by default will import all, and you can change it to
> import specified.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/4853/files#r40898234>.
>

soltysh replied:
We've talked yesterday about the implicit policy, the idea was use the default and just specify a list of tags you want to import without the need to specify the underlying image each time. But your solution sounds ok to me, unless @ncdc agrees as well, since he was voicing the need to have implicit import policy.

smarterclayton replied:
Today we have:

1. Spec.DockerImageRepository
  - if you set this, everything is imported

We are proposing to add:

2. Automatic import of Spec.Tag.Kind == DockerImage

You're describing:

3. Spec.DockerImageRepository + List of tags to import

However, we already can do that with 2 (Tag[0] = "image/foo:bar", Tag[1] =
"image/foo:bar2").  So I don't think we should add another way to do that.


On Thu, Oct 1, 2015 at 2:22 PM, Maciej Szulik <notifications@github.com>
wrote:

> In pkg/image/controller/controller.go
> <https://github.com/openshift/origin/pull/4853#discussion_r40906622>:
>
> > @@ -39,35 +33,105 @@ const retryCount = 2
> >  // the image stream is not modified (so it will be tried again later). If a permanent
> >  // failure occurs the image is marked with an annotation. The tags of the original spec image
> >  // are left as is (those are updated through status).
> > +// There are 3 use cases here:
> > +// 1. spec.DockerImageRepository defined without any tags results in all tags being imported
> > +//    from default registry
> > +// 2. spec.DockerImageRepository + tags defined - import only the provided tags,
> > +//    for kind == DockerImage import underlying image, iow. exact tag (no provided means latest),
> > +//    for kind != DockerImage import tag from the default registry
>
> We've talked yesterday about the implicit policy, the idea was use the
> default and just specify a list of tags you want to import without the need
> to specify the underlying image each time. But your solution sounds ok to
> me, unless @ncdc <https://github.com/ncdc> agrees as well, since he was
> voicing the need to have implicit import policy.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/4853/files#r40906622>.
>

soltysh replied:
Ok, I feel convinced ;)

smarterclayton commented on pkg/image/controller/controller.go:
```diff
@@ -39,35 +33,105 @@ const retryCount = 2
 	}
+
 	name := stream.Spec.DockerImageRepository
+	insecure := stream.Annotations != nil && stream.Annotations[api.InsecureRepositoryAnnotation] == "true"
```
You don't need to check for nil - retrieving an item from a nil map returns `<default value for map value type>, false`

miminar replied:
Shouldn't the comparison be case-insensitive?

soltysh replied:
It is we who set that annotation, so no atm.

ncdc commented on pkg/image/registry/imagestreammapping/rest.go:
```diff
@@ -69,6 +70,8 @@ func (s *REST) Create(ctx kapi.Context, obj runtime.Object) (runtime.Object, err
 
 	mapping := obj.(*api.ImageStreamMapping)
 
+	glog.Infof("Creating mapping %s %s %s", mapping.Name, mapping.Tag, mapping.Image.DockerImageReference)
```
V(2)? or something other than on by default

soltysh replied:
That's a debug leftover, sorry for that.

miminar commented on pkg/image/controller/controller.go:
```diff
@@ -39,35 +33,105 @@ const retryCount = 2
 		err = fmt.Errorf("invalid docker image repository, cannot import data: %v", err)
-		util.HandleError(err)
-		return c.done(stream, err.Error(), retryCount)
+		return tags, err, true
```
Maybe I miss something, but does it make sense to retry string parsing if the `name` doesn't change (which I assume)?

soltysh replied:
The same applies as with the following questions of yours, @ncdc can you please chime in?

miminar commented on pkg/image/controller/controller.go:
```diff
@@ -39,35 +33,105 @@ const retryCount = 2
 	conn, err := client.Connect(ref.Registry, insecure)
 	if err != nil {
-		return err
+		return tags, err, false
```
I believe this should be retried.

soltysh replied:
I didn't intend to change the current logic, so that's a question to @ncdc rather. On the other hand I can agree that retrying makes sense, unless there's something I don't know about connecting to image registry.

ncdc replied:
I don't see any harm in retrying here.

miminar commented on pkg/image/controller/controller.go:
```diff
@@ -39,35 +33,105 @@ const retryCount = 2
+}
+
+// importDefaultTags imports tags from the stream's default registry
+// returns an error and bool flag specifying if the error is retryable
```
`s/retryable/retriable/`

soltysh replied:
After googling a while I didn't find a reasonable result one vs the other. But our sources prefer the version with 'y' :wink:

miminar replied:
LOL. I'm not a native speaker so I won't argue with you. It just looked funny to me.
And [thesaurus](http://www.thesaurus.com/misspelling?term=retryable&s=t) got me no results (though `retriable` is known).

Michal
On 1.10.2015 10:30, Maciej Szulik wrote:
> In pkg/image/controller/controller.go <https://github.com/openshift/origin/pull/4853#discussion_r40889661>:
>
>>  	}
>> -	tags, err := conn.ImageTags(ref.Namespace, ref.Name)
>> -	switch {
>> -	case dockerregistry.IsRepositoryNotFound(err), dockerregistry.IsRegistryNotFound(err):
>> -		return c.done(stream, err.Error(), retryCount)
>> -	case err != nil:
>> -		return err
>> +	tags, err = conn.ImageTags(ref.Namespace, ref.Name)
>> +
>> +	return tags, err, dockerregistry.IsRepositoryNotFound(err) || dockerregistry.IsRegistryNotFound(err)
>> +}
>> +
>> +// importDefaultTags imports tags from the stream's default registry
>> +// returns an error and bool flag specifying if the error is retryable
>
> After googling a while I didn't find a reasonable result one vs the other. But our sources prefer the version with 'y'
> :wink:
>
> —
> Reply to this email directly or view it on GitHub <https://github.com/openshift/origin/pull/4853/files#r40889661>.
>

soltysh commented:
I'm done with the implementation, @ncdc ptal.

ncdc commented on pkg/image/controller/controller.go:
```diff
@@ -39,20 +33,77 @@ const retryCount = 2
+// 2. spec.DockerImageRepository + tags defined - import all tags from default registry,
+//    and all the specified which (if name matches) will overwrite the default ones.
+//    Additionally:
+//    for kind == DockerImage import or reference underlying image, iow. exact tag (no provided means latest),
```
s/no provided/not provided/

ncdc commented on pkg/image/controller/controller.go:
```diff
@@ -39,20 +33,77 @@ const retryCount = 2
 // are left as is (those are updated through status).
+// There are 3 use cases here:
+// 1. spec.DockerImageRepository defined without any tags results in all tags being imported
+//    from default registry
```
s/default registry/upstream image repository/

ncdc commented on pkg/image/controller/controller.go:
```diff
@@ -39,20 +33,77 @@ const retryCount = 2
+// There are 3 use cases here:
+// 1. spec.DockerImageRepository defined without any tags results in all tags being imported
+//    from default registry
+// 2. spec.DockerImageRepository + tags defined - import all tags from default registry,
```
s/default registry/upstream image repository/

ncdc commented on pkg/image/controller/controller.go:
```diff
@@ -39,20 +33,77 @@ const retryCount = 2
 	}
 
-	insecure := stream.Annotations != nil && stream.Annotations[api.InsecureRepositoryAnnotation] == "true"
+	importTags := make(map[string]string)
```
This is an unnecessary allocation if you hit line 63

ncdc commented on pkg/image/controller/controller.go:
```diff
@@ -39,20 +33,77 @@ const retryCount = 2
+	var tags map[string]string
+	name := stream.Spec.DockerImageRepository
+	if len(name) == 0 {
+		// no default image specified, or no tags to import
```
no default image repository defined (remove bit about no tags)

ncdc commented on pkg/image/controller/controller.go:
```diff
@@ -39,20 +33,77 @@ const retryCount = 2
+	ref, err := api.ParseDockerImageReference(name)
+	if err != nil {
+		err = fmt.Errorf("invalid docker image repository, cannot import data: %v", err)
+		return tags, err, true
```
this doesn't seem retryable

soltysh commented:
@ncdc comments addressed and reorganized the code per your suggestion. ptal.

[test] in the mean time.

smarterclayton commented on pkg/image/api/v1beta3/types.go:
```diff
@@ -63,6 +63,8 @@ type NamedTagReference struct {
 	Annotations map[string]string     `json:"annotations,omitempty"`
 	From        *kapi.ObjectReference `json:"from,omitempty"`
+	// Referenace states if the tag will be imported. Default value is false, which means the tag will be imported.
+	Reference bool `json:"reference,omitempty" description:"flag informing if tag is a reference (true) or is imported"`
```
`if true consider this tag a reference only and do not attempt to import metadata about the image`

soltysh replied:
done.

smarterclayton commented on pkg/image/controller/controller.go:
```diff
@@ -22,78 +24,146 @@ type ImportController struct {
+			if specTag.From.Kind == "DockerImage" && !specTag.Reference {
+				ref, err := api.ParseDockerImageReference(specTag.From.Name)
+				if err != nil {
+					glog.V(2).Infof("error parsing DokcerImage %s: %v", specTag.From.Name, err)
```
Typo

soltysh replied:
done.

smarterclayton commented on pkg/image/controller/controller.go:
```diff
@@ -22,78 +24,146 @@ type ImportController struct {
+	// read explicitly defined tags
+	for tagName, specTag := range stream.Spec.Tags {
+		if specTag.From != nil {
+			if specTag.From.Kind == "DockerImage" && !specTag.Reference {
```
break the nesting here:

```
if specTag.From == nil {
  continue
}
if specTag.From.Kind != "DockerImage" || specTag.Reference {
  references.Insert(tagName)
  continue
}
...
```

soltysh replied:
done.

smarterclayton commented on pkg/image/controller/controller.go:
```diff
@@ -22,78 +24,146 @@ type ImportController struct {
+		if image, ok := retrieved[ref.ID]; ok {
+			dockerImage = image
+		} else {
+			// TODO insecure applies to the stream's spec.dockerImageRepository, not necessarily to an external one!
```
Actually, insecure can be seen as applying to all import locations.

ncdc replied:
We may need to add another insecure flag per spec tag. I don't think we should use the image stream's insecure annotation to apply to individual spec tag refs.

smarterclayton replied:
Why?

On Tue, Oct 6, 2015 at 2:50 PM, Andy Goldstein <notifications@github.com>
wrote:

> In pkg/image/controller/controller.go
> <https://github.com/openshift/origin/pull/4853#discussion_r41305797>:
>
> > -		switch {
> > -		case dockerregistry.IsRepositoryNotFound(err), dockerregistry.IsRegistryNotFound(err):
> > -			return c.done(stream, err.Error(), retryCount)
> > -		case dockerregistry.IsImageNotFound(err):
> > -			continue
> > -		case err != nil:
> > -			return err
> > +// importTags imports tags specified in a map from given ImageStream. Returns an error if one occurs.
> > +func (c *ImportController) importTags(stream *api.ImageStream, imports map[string]api.DockerImageReference, client dockerregistry.Client, insecure bool) error {
> > +	retrieved := make(map[string]*dockerregistry.Image)
> > +	for tag, ref := range imports {
> > +		var dockerImage *dockerregistry.Image
> > +		if image, ok := retrieved[ref.ID]; ok {
> > +			dockerImage = image
> > +		} else {
> > +			// TODO insecure applies to the stream's spec.dockerImageRepository, not necessarily to an external one!
>
> We may need to add another insecure flag per spec tag. I don't think we
> should use the image stream's insecure annotation to apply to individual
> spec tag refs.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/4853/files#r41305797>.
>

ncdc replied:
I guess it's not critical. However, users could be confused if they add the insecure annotation to the stream and `--insecure-registry` to `/etc/sysconfig/docker` for the registry/stream in question, but they don't have it for a spec tag ref that pulls from another insecure registry that's not listed in `/etc/sysconfig/docker`. Anyways, all registries should be secured ( :smile: ), so ...

soltysh replied:
As @ncdc stated, let's not forget that tag can now import data from registry other than the default one.

ncdc commented:
I will review this tomorrow morning when I'm back from PTO.

smarterclayton commented on pkg/image/controller/controller.go:
```diff
@@ -22,78 +24,146 @@ type ImportController struct {
+				dockerImage, err = conn.ImageByTag(ref.Namespace, ref.Name, ref.Tag)
+			}
+			switch {
+			case dockerregistry.IsRepositoryNotFound(err), dockerregistry.IsRegistryNotFound(err):
```
This is no longer the case where it should terminate an import.  Instead, you need to make as much progress as possible and then fail at the end.  It would be better to accumulate errors and continue.  `c.done()` should not be invoked in this loop.  This iteration of the loop should be extracted into a function.

soltysh replied:
We don't have any reasonable error handling for IS. Do we want to use
events for that or you're talking about just logging errors here? I agree
we should process as much as possible, but still we want to inform user
about import problems in a reasonable way.
On Oct 6, 2015 8:51 PM, "Clayton Coleman" <notifications@github.com> wrote:

> In pkg/image/controller/controller.go
> <https://github.com/openshift/origin/pull/4853#discussion_r41306035>:
>
> > +		var dockerImage *dockerregistry.Image
> > +		if image, ok := retrieved[ref.ID]; ok {
> > +			dockerImage = image
> > +		} else {
> > +			// TODO insecure applies to the stream's spec.dockerImageRepository, not necessarily to an external one!
> > +			conn, err := client.Connect(ref.Registry, insecure)
> > +			if err != nil {
> > +				return err
> > +			}
> > +			if len(ref.ID) > 0 {
> > +				dockerImage, err = conn.ImageByID(ref.Namespace, ref.Name, ref.ID)
> > +			} else {
> > +				dockerImage, err = conn.ImageByTag(ref.Namespace, ref.Name, ref.Tag)
> > +			}
> > +			switch {
> > +			case dockerregistry.IsRepositoryNotFound(err), dockerregistry.IsRegistryNotFound(err):
>
> This is no longer the case where it should terminate an import. Instead,
> you need to make as much progress as possible and then fail at the end. It
> would be better to accumulate errors and continue. c.done() should not be
> invoked in this loop. This iteration of the loop should be extracted into a
> function.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/4853/files#r41306035>.
>

smarterclayton replied:
We do, it goes in the annotation.  But you don't even need to do more than
report the first error.


On Tue, Oct 6, 2015 at 5:36 PM, Maciej Szulik <notifications@github.com>
wrote:

> In pkg/image/controller/controller.go
> <https://github.com/openshift/origin/pull/4853#discussion_r41325904>:
>
> > +		var dockerImage *dockerregistry.Image
> > +		if image, ok := retrieved[ref.ID]; ok {
> > +			dockerImage = image
> > +		} else {
> > +			// TODO insecure applies to the stream's spec.dockerImageRepository, not necessarily to an external one!
> > +			conn, err := client.Connect(ref.Registry, insecure)
> > +			if err != nil {
> > +				return err
> > +			}
> > +			if len(ref.ID) > 0 {
> > +				dockerImage, err = conn.ImageByID(ref.Namespace, ref.Name, ref.ID)
> > +			} else {
> > +				dockerImage, err = conn.ImageByTag(ref.Namespace, ref.Name, ref.Tag)
> > +			}
> > +			switch {
> > +			case dockerregistry.IsRepositoryNotFound(err), dockerregistry.IsRegistryNotFound(err):
>
> We don't have any reasonable error handling for IS. Do we want to use
> events for that or you're talking about just logging errors here? I agree
> we should process as much as possible, but still we want to inform user
> about import problems in a reasonable way.
> … <#1503f1535551d805_>
> On Oct 6, 2015 8:51 PM, "Clayton Coleman" <notifications@github.com>
> wrote: In pkg/image/controller/controller.go <
> https://github.com/openshift/origin/pull/4853#discussion_r41306035>: > +
> var dockerImage *dockerregistry.Image > + if image, ok :=
> retrieved[ref.ID]; ok { > + dockerImage = image > + } else { > + // TODO
> insecure applies to the stream's spec.dockerImageRepository, not
> necessarily to an external one! > + conn, err :=
> client.Connect(ref.Registry, insecure) > + if err != nil { > + return err >
> + } > + if len(ref.ID) > 0 { > + dockerImage, err =
> conn.ImageByID(ref.Namespace, ref.Name, ref.ID) > + } else { > +
> dockerImage, err = conn.ImageByTag(ref.Namespace, ref.Name, ref.Tag) > + }
> > + switch { > + case dockerregistry.IsRepositoryNotFound(err),
> dockerregistry.IsRegistryNotFound(err): This is no longer the case where it
> should terminate an import. Instead, you need to make as much progress as
> possible and then fail at the end. It would be better to accumulate errors
> and continue. c.done() should not be invoked in this loop. This iteration
> of the loop should be extracted into a function. — Reply to this email
> directly or view it on GitHub <
> https://github.com/openshift/origin/pull/4853/files#r41306035>.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/4853/files#r41325904>.
>

kargakis commented:
Is this ready for review @soltysh ?

soltysh commented:
It is. Sorry I forgot to tag you here :-/
On Oct 6, 2015 11:18 PM, "Michail Kargakis" <notifications@github.com>
wrote:

> Is this ready for review @soltysh <https://github.com/soltysh> ?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/4853#issuecomment-146004608>.
>

kargakis commented on pkg/image/controller/controller.go:
```diff
@@ -22,78 +24,146 @@ type ImportController struct {
 
 // Next processes the given image stream, looking for streams that have DockerImageRepository
-// set but have not yet been marked as "ready". If transient errors occur, err is returned but
+// set but have not yet been marked as "ready". If transient errors occurs, err is returned but
```
I think "occur" is correct

soltysh replied:
yup - fixed.

kargakis commented on pkg/image/controller/controller.go:
```diff
@@ -22,78 +24,146 @@ type ImportController struct {
 	for tag, image := range tags {
-		if specTag, ok := stream.Spec.Tags[tag]; ok && specTag.From != nil {
-			// spec tag is set to track another tag - do not import
+		if _, ok := imports[tag]; ok {
```
```go
if _, ok := imports[tag]; ok || references.Has(tag) {
```

soltysh replied:
done.

kargakis commented on pkg/image/controller/controller.go:
```diff
@@ -22,78 +24,146 @@ type ImportController struct {
+			if t2 == tag {
+				continue
+			}
+			if i2 == image && t2 == image {
```
Check first for `t2 == image` and you won't need the `t2 == tag` above at all

soltysh replied:
done.

kargakis commented on pkg/image/controller/controller_test.go:
```diff
@@ -85,6 +87,168 @@ func TestControllerNoDockerRepo(t *testing.T) {
+	if err := c.Next(&stream); err != nil {
+		t.Errorf("unexpected error: %v", err)
+	}
+	if len(stream.Annotations["openshift.io/image.dockerRepositoryCheck"]) == 0 {
```
Use the constant we have for this

kargakis replied:
You are using it likewise in a couple of other places in the tests. Update them too

soltysh replied:
done.

soltysh commented:
@smarterclayton I've addressed your comments, changing the logic so that it goes as far as it can without actually throwing an error, but with this solution the question is does it mean we should retry until there are no errors? Current implementation won't retry because of not returning those errors at all (just in annotation), but that's why I'm asking. Additionally is there some limit on annotation length? I'm currently doing error aggregate and then transform that into string which goes into annotation, but I'm positive this can lean to overflowing eventual annotation limit length, so I'd appreciate info.

smarterclayton commented:
On Oct 9, 2015, at 6:21 AM, Maciej Szulik <notifications@github.com> wrote:

@smarterclayton <https://github.com/smarterclayton> I've addressed your
comments, changing the logic so that it goes as far as it can without
actually throwing an error, but with this solution the question is does it
mean we should retry until there are no errors? Current implementation
won't retry because of not returning those errors at all (just in
annotation), but that's why I'm asking.



The error behavior should mimic the old state - certain errors guarantee
failure, others do not.



Additionally is there some limit on annotation length? I'm currently doing
error aggregate and then transform that into string which goes into
annotation, but I'm positive this can lean to overflowing eventual
annotation limit length, so I'd appreciate info.



You should trim the error to some max length (300 chars or so) before
saving it.


—
Reply to this email directly or view it on GitHub
<https://github.com/openshift/origin/pull/4853#issuecomment-146825084>.

soltysh commented:
@kargakis @smarterclayton @ncdc this is ready for final review. 
I've kept backwards compatibility by returning pretty much the same errors when we fail, but still we'll try to reach as far as possible when actually importing. The import logic is also backward compatible with previously defined image streams.

ncdc commented on pkg/image/api/v1/types.go:
```diff
@@ -68,6 +68,8 @@ type NamedTagReference struct {
 	Annotations map[string]string `json:"annotations,omitempty" description:"annotations associated with images using this tag"`
 	// From is a reference to an image stream tag or image stream this tag should track
 	From *kapi.ObjectReference `json:"from,omitempty" description:"a reference to an image stream tag or image stream this tag should track"`
+	// Referenace states if the tag will be imported. Default value is false, which means the tag will be imported.
```
typo: Referenace

ncdc commented on pkg/image/api/types.go:
```diff
@@ -79,6 +79,8 @@ type TagReference struct {
 	Annotations map[string]string
 	// Optional; if specified, a reference to another image that this tag should point to. Valid values are ImageStreamTag, ImageStreamImage, and DockerImage.
 	From *kapi.ObjectReference
+	// Referenace states if the tag will be imported. Default value is false, which means the tag will be imported.
```
Typo

ncdc commented:
@pweil- @miminar

miminar commented on pkg/image/controller/controller.go:
```diff
@@ -39,126 +37,203 @@ const retryCount = 2
+// getTags returns tags from default upstream image repository and explicitly defined.
+// Returns a map of tags to be imported and an error if one occurs.
+// Tags explicitly defined will overwrite those from default upstream image repository.
+func (c *ImportController) getTags(stream *api.ImageStream, client dockerregistry.Client, insecure bool) (map[string]api.DockerImageReference, error) {
```
nitpick: Controller is unused inside this method. It could be turned into a function.

soltysh replied:
Good catch, it used to be, but not anymore.

miminar commented on pkg/image/controller/controller_test.go:
```diff
@@ -66,14 +68,15 @@ func (f *fakeDockerRegistryClient) ImageByID(namespace, name, id string) (*docke
+			Namespace:   "other",
 		},
 	}
 	other := stream
```
I believe that simple assignment doesn't achieve desired conditions. Both `other` and `stream` will point to the same `Annotations` map.

soltysh replied:
Yeah, I am worried with that as well. I'll fix it when addressing comments tomorrow.

soltysh commented:
Typos from @ncdc and nits from @miminar addressed.

miminar commented:
LGTM

kargakis commented on pkg/image/controller/controller.go:
```diff
@@ -39,126 +37,203 @@ const retryCount = 2
-		for _, tag := range tags {
-			if idTagPresent && id == tag {
-				continue
+		for t2, i2 := range tags {
```
It's not really obvious to me what's happening here, mind adding a comment?

soltysh commented:
@kargakis I've added there a comment about keeping backwards compatibility with v1 registries, which did not provide imageID info with tag. Is that enough for you?

kargakis commented:
> @kargakis I've added there a comment about keeping backwards compatibility with v1 registries, which did not provide imageID info with tag. Is that enough for you?

:+1: 

LGTM

soltysh commented:
Got 2.
[merge]
On Oct 15, 2015 11:17 PM, "OpenShift Bot" <notifications@github.com> wrote:

> Evaluated for origin test up to 04c69ef
> <https://github.com/openshift/origin/commit/04c69efc19141dc9204a1d2d803908dd43582133>
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/4853#issuecomment-148523621>.
>

soltysh commented:
I just rerun the failing tests locally - works as expected. In the mean time I saw the job was picked up by jenkins. I'll wait for the result and then remerge again if if won't do that before.

soltysh commented:
I'm closing and reopening to kick travis, I'll also kick the jenkins to have it retested. All of the failures are passing on my local machine.

soltysh commented:
re-[test]

soltysh commented:
This seems to fail only with `hack/test-end-to-end-docker.sh`, the local version works as expected. Let's give a jenkins another try.

soltysh commented:
I had to use `ParseDockerImageReference` instead of the initial `DockerImageReferenceForStream` in `getTags` method. `ParseDockerImageReference` was used in the original implementation, whereas the other one came from a discussion with @ncdc.

soltysh commented:
re-[test] the errors are flake, I can't repro locally.

smarterclayton commented:
The test failure in hack/test-cmd.sh may be your changes.  Can you run it on a loop on your laptop and verify it completes?  Did not see that failure before your changes.

soltysh commented:
@smarterclayton I can't reproduce the error from `hack/test-cmd.sh`, it's working ok, I'll try to run that in the loop as you proposed, though. What worries me more is the `hack/test-end-to-end-docker.sh` which runs perfectly ok as `hack/test-end-to-end.sh` but the docker version fails on AWS. I'm working on it... stilll...

soltysh commented:
@smarterclayton I figured out what is causing this problem. Due to my changes in the import the time needed for importing the image is a bit longer than previously. This infects the flow in [registry](https://github.com/openshift/origin/blob/master/pkg/dockerregistry/server/repositorymiddleware.go#L200-L209), where if IS is missing upon ISM creation we create new IS and right after that ISM, which is getting version mismatch error: `Error creating image stream mapping: imageStream \"ruby-20-centos7\" cannot be updated: the
 object has been modified; please apply your changes to the latest version and try again`. I've talked with @miminar and @ncdc I'm working on adding a retry to the ISM creation, that should solve our problem.

soltysh commented:
I'm waiting for #5301 to land and then I'll rebase to catch the solution to failing tests in this PR.

soltysh commented:
Rebased to pick up the solution to failing tests.

soltysh commented:
Finally, :facepunch: upon having 2 LGTMs:
[merge]

pweil- commented:
:tada: :balloon: :cake:
